blog: "Stylo - Another Kind Of Rich Text Editor" (#66)
peterpeterparker/daviddalbusco.com: src/blog/stylo-another-kind-of-rich-text-editor.md

---
path: "/blog/stylo-another-kind-of-rich-text-editor"
date: "2022-02-08"
title: "Stylo - Another Kind Of Rich Text Editor"
description: "So, I developed a new open source WYSIWYG interactive editor for JavaScript."
tags: "#javascript #showdev #webdev #webcomponents"
image: "https://cdn-images-1.medium.com/max/1600/1*kPGH8RB1reGGiq6nU0UaIg.jpeg"
canonical: "https://zhx6p-eqaaa-aaaai-abbrq-cai.raw.ic0.app/d/converting-svg-to-image-in-javascript"
---

![](https://cdn-images-1.medium.com/max/1600/1*kPGH8RB1reGGiq6nU0UaIg.jpeg)

*Photo by [Amy Hirschi](https://unsplash.com/@amyhirschi?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText) on [Unsplash](https://unsplash.com/s/photos/pen?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText)*

*****

So, I developed a new open source WYSIWYG interactive editor for JavaScript. Its goal is to bring great user experience and interactivity to the web, for everyone, with no dependencies. I called it: [Stylo](https://stylojs.com/).

*****

### But Why?

> While developing last year the migration of our web editor for slides, [DeckDeckGo](https://deckdeckgo.com), to [DFINITY](https://dfinity.org)’s Internet Computer, we also initialized a new experimental concept to support documents as well.

Reviewer: ```suggestion
Last year, while migrating our web editor for slides, [DeckDeckGo](https://deckdeckgo.com), to [DFINITY](https://dfinity.org)’s Internet Computer, we started playing around with some new code to support support documents as well.
```

> As a very first test, I wrote and published a blog post on our platform — [Converting SVG To Image (PNG, Webp, etc.) in JavaScript](https://zhx6p-eqaaa-aaaai-abbrq-cai.raw.ic0.app/d/converting-svg-to-image-in-javascript).

Reviewer: Do you mean that converting SVG to image was the first test, or that you wrote the article using the new tool?

Author: reference removed, makes clearer.

> Although I was satisfied with the result, I was of course also aware that further enhancements were needed.

Reviewer: ```suggestion
Although I was satisfied with the result, I thought: this can be better.
```

> That’s why I had the idea to extract the rich text editor capabilities of the core of our application. By limiting the scope of these features, I assumed it would be easier to improve and fix what still needed to be done.

Reviewer: ```suggestion
That’s why I had the idea to extract the rich text editor capabilities of the core of our application -- under the assumption that it would be easier to improve and fix the text editor in isolation.
```

> In addition, by publishing the new library as a new open source repo, I was also hoping it would benefit from feedbacks and contributions.

Reviewer: ```suggestion
In addition, by publishing the new library as a new open source repo, I was also hoping it would benefit from community feedback and contributions.
```

What can I add? I am an engineer, I like to code 😉.

*****

### Features

Not looking to oversell what Stylo can do — or do differently — but here are in my humble opinion a list of cool things it can do:

*****

#### Interactive Design 🎯

> Unlike other WYSIWYG editor Stylo has no sticky toolbar. It is inspired by design and user experience of platforms that I like — i.e [Notion](https://www.notion.so/) and [Medium](https://medium.com/).

Reviewer: ```suggestion
Unlike other WYSIWYG editors, Stylo has no sticky toolbar. It is inspired by design and user experience of platforms that I like — i.e [Notion](https://www.notion.so/) and [Medium](https://medium.com/).
```

![](https://cdn-images-1.medium.com/max/1600/1*D-YYgIa5cY0Ya8FAN0Lyiw.gif)

> When writing the document, user can use different types of paragraph that can be added through a floating menu. These elements can be simple HTML elements — i.e. h1, h2, code etc. — or more complex block — i.e. any Web Components.

Reviewer: ```suggestion
When writing the document, the user can use different types of paragraph that can be added through a floating menu. These elements can be simple HTML elements — i.e. h1, h2, code etc. — or more complex block — i.e. any Web Component.
```

> For styling purpose, another toolbar gives the user the ability to modify text — e.g. bold, italic, colors, etc. — and add hyperlinks.

Reviewer: ```suggestion
For styling purposes, another toolbar gives the user the ability to modify text — e.g. bold, italic, colors, etc. — and add hyperlinks.
```

> Last but not least, for an optimized experience on desktop, many actions can be performed without the need of the mouse. Opening the list of paragraphs can be triggered hitting “/” and **bold**, *italic* and `mark` are accessible through keyboard shortcuts.

Reviewer: ```suggestion
Last but not least, for an optimized experience on desktop, many actions can be performed without the mouse. Opening the list of paragraphs can be triggered hitting “/” and **bold**, *italic* and `mark` are accessible through keyboard shortcuts.
```

Reviewer: note: I had never heard of `mark` until stylo

Author: should it be another element?

*****

#### Customizable 💪

Conceptually, only one main element — a container — has to be set as editable and all its children are either HTML Elements or Web Components — i.e. no direct text node descendant.

> This decision was made for performance reason. In the future we might need to lazy load the content or improve the scrolling experience with some sort of infinite or other virtual scrolling approach. By limiting children to elements, it is possible to assign attributes such as unique ids directly in the DOM. <br>

Reviewer: ```suggestion
This decision was made for performance reasons. In the future we might need to lazy load the content or improve the scrolling experience with some sort of infinite scrolling or some other virtual scrolling approach. By limiting children to elements, it is possible to assign attributes such as unique ids directly in the DOM.
```

\<Brrrr> 🥶

> This brings us to the customization. The list of paragraphs that can be added by the user is fully customizable. One entry — i.e. a plugin — in the list is a text, an icon and a function that adds a new element to the editable container.

Reviewer: > One entry — i.e. a plugin — in the list is a text, an icon and a function that adds a new element to the editable container.

I'm struggling to understand this. Do you mean that by hitting e.g. `/`, the user can then... add a function to the document?

Reviewer: Ok, looking at the snippet below it makes more sense. I think this is the part that threw me off:

> The list of paragraphs that can be added by the user is fully customizable.

I thought you meant "the list of HTML elements in the document", i.e. the paragraphs themselves, whereas you actually mean the type of paragraphs. Correct?

Author: rephrased as following, hope it is a bit clearer:

The miscellaneous types of paragraphs that can be added by the user - e.g. block of code, titles, etc. - is fully customizable. One type - i.e. a plugin - is defined with a ​text​, an ​icon​ and a​ function that adds a new element to the document.

> For example, if we would implement a plugin that generates a new paragraph that is itself a Web Component name <hello-world/>, the solution would look as following:

Reviewer: ```suggestion
For example, if we were to implement a plugin that generates a new paragraph that is itself a Web Component name <hello-world/>, the solution would look something like this:
```

```javascript
import {createEmptyElement, StyloPlugin, StyloPluginCreateParagraphsParams, transformParagraph} from '@papyrs/stylo';
export const hr: StyloPlugin = {
  text: 'My Hello World',
  icon: `<svg width="32" height="32" viewBox="0 0 512 512">
        ...
    </svg>
  `,
  createParagraphs: async ({container, paragraph}: StyloPluginCreateParagraphsParams) => {
    // Create your Web Component or HTML Element
    const helloWorld = document.createElement('hello-world');
    // Set properties, attributes or styles
    helloWorld.setAttributes('yolo', 'true');
     transformParagraph({
       elements: [helloWorld, createEmptyElement({nodeName: 'div'})],
       paragraph,
       container,
       focus: 'first'
     })
  }
};
```

> Stylo exposes interfaces and utilities to ease the development of custom plugins. The most important being maybe transformParagraph which parses the elements to the DOM.

Reviewer: ```suggestion
Stylo exposes interfaces and utilities to ease the development of custom plugins. The most important being maybe `transformParagraph` which parses the elements to the DOM.
```

What do you mean "parses the element **to** the DOM"?

> Finally, it is worth to note the function createParagraphs is async which means the process of creating new elements can totally be hijacked by your application. As for example, when integrated in DeckDeckGo, the function does not always directly create new elements. For some more complex use cases, when the function is triggered, it opens modals that require more user interactions such as selecting an image before effectively modifying the content.

Reviewer: ```suggestion
Finally, it is worth noting that the function `createParagraphs` is async which means the process of creating new elements can totally be hijacked by your application. As for example, when integrated in DeckDeckGo, the function does not always directly create new elements. For some more complex use cases, when the function is triggered, it opens modals that require more user interactions such as selecting an image before effectively modifying the content.
```

❤️

![](https://cdn-images-1.medium.com/max/1600/1*9GgkGbx_4D-hIwT8PxJivQ.gif)

*****

#### Framework Agnostic 😎

Stylo is a framework agnostic library of reusable components developed with [StencilJS](https://stenciljs.com/). It works on all modern browsers.

*****

#### Lightweight 🪶

> 30kb gzipped might in certain circumstances already sound like a lot, specially to me, but in terms of WYSIWYG editor it actually turns out be lightweight.

Reviewer: ```suggestion
30kb gzipped might in certain circumstances already sound like a lot, especially to me, but in terms of WYSIWYG editor it actually turns out be lightweight.
```

> From what I observed, it might even be the lightest. There is also probably a bit of space for improvements.

Reviewer: ```suggestion
From what I observed, it might even be the lightest. There is also probably room for improvement.
```

*****

#### Future Proof 🚀

> While a handful of actions in the alpha version still rely on execCommand to apply styles (e.g. bold, italic) — the core of the library does not.

Reviewer: ```suggestion
While a handful of actions in the alpha version still rely on `execCommand` to apply styles (e.g. bold, italic) — the core of the library does not.
```

Been adding those backticks, but actually maybe you don't `mark` the words on purpose? From now on I'll refrain from adding backticks

> Stylo controls what happens on every mutation. It uses the MutationObserver API to keep track of the changes and stack them in a custom “undo-redo” implementation. It also forward the information about what changed in the DOM to your application with custom events.

Reviewer: ```suggestion
Stylo controls what happens on every mutation. It uses the MutationObserver API to keep track of the changes and stacks them in a custom “undo-redo” implementation. It also forwards the information about what changed in the DOM to your application with custom events.
```

Plugins are applied using plain vanilla JavaScript and the Selection API is also used to detect and apply transformations.

> That’s why, regardless of what will happen to the deprecated API, Stylo is built for the future.

Reviewer: ```suggestion
That's why, regardless of what happens to the deprecated API, Stylo is built for the future.
```

*****

#### Open Source ⭐️

The components are licensed under MIT license. They can be used by anyone, anywhere where the web is used and this even in a closed source application.

To contribute and follow the adventure, watch and star the [GitHub repo](https://github.com/papyrs/stylo).

*****

### Is It Production Ready?

![](https://cdn-images-1.medium.com/max/1600/1*snbd7Pn0GcdqaEa9m48cIQ.gif)

Stylo is currently an **alpha** version! ⚠️ The project is under active development and contributions on GitHub are most welcome.

I did use it — not without noticing many bugs I need to fix — to write this article. It might probably not be ready tomorrow nor the day after but I am confident that it will be ready to start the beta phase of our project that will run on the Internet Computer.

*****

### Conclusion

> I am well aware there is still quite some work to make [Stylo](https://stylojs.com) bulletproof but it is quite exciting because challenging 😉.

Reviewer: ```suggestion
I am well aware there is still quite some work to make [Stylo](https://stylojs.com) bulletproof but it is quite exciting because it is also challenging 😉.
```

> To infinity and beyond!

Reviewer: ```suggestion
To dfinity and beyond!
```

lol jk

Author: 😜

This decision was made for performance reasons. In the future we might need to lazy load the content or improve the scrolling experience with some sort of infinite scrolling or some other virtual scrolling approach. By limiting children to elements, it is possible to assign attributes such as unique ids directly in the DOM.

> This brings us to the customization. The miscellaneous types of paragraphs that can be added by the user - e.g. block of code, titles, etc. - is fully customizable. One type - i.e. a plugin - is defined with a ​text​, an ​icon​ and a​ function that adds a new element to the document.

Reviewer: ```suggestion
This brings us to the customization. The various types of paragraphs that can be added by the user - e.g. block of code, titles, etc. - is fully customizable. Each type - i.e. plugin - is defined with a ​text​, an ​icon​ and a​ function that adds a new element to the document.
```
